### DISCUSSION TITLE
Request for feedback: Parsing Advent of Code

### DISCUSSION THREAD

**Discussion by nicolecomputer** [alecthomas/participle#399] [Show and tell]
Hello!

Thank you for this parser! I am just learning how to use it and I wanted some feedback on a recent practice I did.

I tried [parse an advent of code problem's input](https://github.com/nicolecomputer/go-practice/tree/main/parse-aoc-23-02) using participle. I'm not sure I did it quite right 😅 

My code for parsing is here: https://github.com/nicolecomputer/go-practice/blob/main/parse-aoc-23-02/main.go

**A couple of questions:**
- I noticed parsing strings is hard! Sometimes it seems like I want to use `@String` and sometimes I want to use `@Ident`. How do I know when to use each?
- Is there a recommended way to iteratively build up a parser? The game input is something like `Game 1: red 10, blue 3; green 1` and at first I wanted to pull a struct with **just** `Game 1` and `red 10, blue 3; green 1` but I couldn't figure out how to suck up those strings
- Do you know of a general resource for learning the parsing grammar? I spent a bunch of time looking through examples and looking at the go docs and copy+pasting things without super super understanding what I was doing. I got there in the end but I feel like I missed a beat.

I'm really grateful for the care that you've put into the library (just seeing a tutorial linked at the top made me SO happy 💖 💖 💖 )

🍃 Nicole

**Comment by alecthomas** [comment]
Hi 👋, thanks for the kind words.

> I noticed parsing strings is hard! Sometimes it seems like I want to use @String and sometimes I want to use @Ident. How do I know when to use each?

With the default lexer (see [text/scanner](https://pkg.go.dev/text/scanner#pkg-constants) for more detail), `@String` is always a quoted escaped string in the form `"<text>"` and `@Ident` is always an "identifier" as Go sees it. 

> Is there a recommended way to iteratively build up a parser? The game input is something like Game 1: red 10, blue 3; green 1 and at first I wanted to pull a struct with just Game 1 and red 10, blue 3; green 1 but I couldn't figure out how to suck up those strings

The tutorial shows how I would typically build a parser iteratively, however I will usually incrementally build it up and let it fail. The failure tells me that there's still work to do. If you want it to succeed, you can match all remaining tokens with something like `@(~EOF)` (described [here](https://github.com/alecthomas/participle#grammar-syntax)).

> Do you know of a general resource for learning the parsing grammar? I spent a bunch of time looking through examples and looking at the go docs and copy+pasting things without super super understanding what I was doing. I got there in the end but I feel like I missed a beat.

Nothing specific no. The grammar is based on EBNF, so any general parsing resource will usually cover it.

All the best,
Alec
